pkg/kademlia/kademlia.go

```diff
@@ -5,243 +5,198 @@ package kademlia
 
 import (
 	"context"
-	"crypto/rand"
-	"fmt"
-	"log"
-	"net"
-	"strconv"
+	"errors"
 
-	bkad "github.com/coyle/kademlia"
 	"github.com/zeebo/errs"
 
 	"storj.io/storj/pkg/dht"
+	"storj.io/storj/pkg/node"
+	"storj.io/storj/pkg/provider"
 	proto "storj.io/storj/protos/overlay"
 )
 
+const (
+	alpha                       = 5
+	defaultIDLength             = 256
+	defaultBucketSize           = 20
+	defaultReplacementCacheSize = 5
+)
+
 // NodeErr is the class for all errors pertaining to node operations
 var NodeErr = errs.Class("node error")
 
+// BootstrapErr is the class for all errors pertaining to bootstrapping a node
+var BootstrapErr = errs.Class("bootstrap node error")
+
 //TODO: shouldn't default to TCP but not sure what to do yet
 var defaultTransport = proto.NodeTransport_TCP
 
+// NodeNotFound is returned when a lookup can not produce the requested node
+var NodeNotFound = NodeErr.New("node not found")
+
+type lookupOpts struct {
+	amount int
+}
+
 // Kademlia is an implementation of kademlia adhering to the DHT interface.
 type Kademlia struct {
-	routingTable   RoutingTable
+	routingTable   *RoutingTable
 	bootstrapNodes []proto.Node
-	ip             string
-	port           string
+	address        string
 	stun           bool
-	dht            *bkad.DHT
+	nodeClient     node.Client
+	alpha          int
 }
 
 // NewKademlia returns a newly configured Kademlia instance
-func NewKademlia(id dht.NodeID, bootstrapNodes []proto.Node, ip string, port string) (*Kademlia, error) {
-	if port == "" {
-		return nil, NodeErr.New("must specify port in request to NewKademlia")
-	}
-
-	ips, err := net.LookupIP(ip)
+func NewKademlia(id dht.NodeID, bootstrapNodes []proto.Node, address string) (*Kademlia, error) {
+	self := proto.Node{Id: id.String(), Address: &proto.NodeAddress{Address: address}}
+	rt, err := NewRoutingTable(&self, &RoutingOptions{
+		kpath:        "kbucket.db",
+		npath:        "nbucket.db",
+		idLength:     defaultIDLength,
+		bucketSize:   defaultBucketSize,
+		rcBucketSize: defaultReplacementCacheSize,
+	})
 	if err != nil {
-		return nil, err
+		return nil, BootstrapErr.Wrap(err)
 	}
 
-	if len(ips) <= 0 {
-		return nil, errs.New("Invalid IP")
+	for _, v := range bootstrapNodes {
+		ok, err := rt.addNode(&v)
+		if !ok || err != nil {
+			return nil, err
+		}
 	}
 
-	ip = ips[0].String()
+	k := &Kademlia{
+		alpha:          alpha,
+		routingTable:   rt,
+		bootstrapNodes: bootstrapNodes,
+		address:        address,
+		stun:           true,
+	}
 
-	bnodes, err := convertProtoNodes(bootstrapNodes)
+	// TODO(coyle): Not sure what the correct values are and if they should be configurable or not
+	ca, err := provider.NewCA(context.Background(), 12, 4)
 	if err != nil {
 		return nil, err
 	}
-
-	bdht, err := bkad.NewDHT(&bkad.MemoryStore{}, &bkad.Options{
-		ID:             id.Bytes(),
-		IP:             ip,
-		Port:           port,
-		BootstrapNodes: bnodes,
-	})
-
+	identity, err := ca.NewIdentity()
```

Comment:
you should get the identity from the caller too

Reply:
done